homeassistant/components/enocean/__init__.py

```diff
@@ -1,93 +1,67 @@
-SIGNAL_SEND_MESSAGE = "enocean.send_message"
-
 
-def setup(hass, config):
+async def async_setup(hass, config):
     """Set up the EnOcean component."""
-    serial_dev = config[DOMAIN].get(CONF_DEVICE)
-    dongle = EnOceanDongle(hass, serial_dev)
-    hass.data[DATA_ENOCEAN] = dongle
-
-    return True
-
-
-class EnOceanDongle:
-    """Representation of an EnOcean dongle."""
-
-    def __init__(self, hass, ser):
-        """Initialize the EnOcean dongle."""
-
-        self.__communicator = SerialCommunicator(port=ser, callback=self.callback)
-        self.__communicator.start()
-        self.hass = hass
-        self.hass.helpers.dispatcher.dispatcher_connect(
-            SIGNAL_SEND_MESSAGE, self._send_message_callback
+    # support for text-based configuration (legacy)
+    if DOMAIN not in config:
+        return True
+
+    if hass.config_entries.async_entries(DOMAIN):
+        # We can only have one dongle. If there is already one in the config,
+        # there is no need to import the yaml based config.
+        return True
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config[DOMAIN]
         )
+    )
 
-    def _send_message_callback(self, command):
-        """Send a command through the EnOcean dongle."""
-        self.__communicator.send(command)
-
-    def callback(self, packet):
-        """Handle EnOcean device's callback.
-
-        This is the callback function called by python-enocan whenever there
-        is an incoming packet.
-        """
+    return True
 
-        if isinstance(packet, RadioPacket):
-            _LOGGER.debug("Received radio packet: %s", packet)
-            self.hass.helpers.dispatcher.dispatcher_send(SIGNAL_RECEIVE_MESSAGE, packet)
 
+async def async_setup_entry(
+    hass: core.HomeAssistant, config_entry: config_entries.ConfigEntry
+):
+    """Set up an EnOcean dongle for the given entry."""
+    enocean_data = hass.data.setdefault(DATA_ENOCEAN, {})
+    usb_dongle = EnOceanDongle(hass, config_entry.data[CONF_DEVICE])
+    await usb_dongle.async_setup()
+    enocean_data[ENOCEAN_DONGLE] = usb_dongle
 
-class EnOceanDevice(Entity):
-    """Parent class for all devices associated with the EnOcean component."""
+    return True
 
-    def __init__(self, dev_id, dev_name="EnOcean device"):
-        """Initialize the device."""
-        self.dev_id = dev_id
-        self.dev_name = dev_name
 
-    async def async_added_to_hass(self):
-        """Register callbacks."""
-        self.async_on_remove(
-            self.hass.helpers.dispatcher.async_dispatcher_connect(
-                SIGNAL_RECEIVE_MESSAGE, self._message_received_callback
-            )
+async def async_unload_entry(hass, config_entry):
+    """Unload ENOcean config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(config_entry, component)
```

Comment:
I'm not sure if the forward unload will work since we don't forward the entry setup. Have you tried this, also without configuring any platforms?

Reply:
Indeed, I forgot to remove the forwarding of unload.
It is now done.